tests/regression/33-constants/03-empty-not-dead-branch.c

```diff
@@ -0,0 +1,16 @@
+//PARAM: --set ana.activated '["constants"]'
+
+int g;
+
+int main() {
+    // state: {bot}, because no locals
```

Comment:
Is this not some sort of bot and top confusion in the constant analysis here? One would assume that we start out with top and not bot

Reply:
Constants starts with a bottom `MapBot`, which is an empty map with missing values implicitly bottom:
https://github.com/goblint/analyzer/blob/262b6dab31938fafbc3cd9237f40b134812bd27d/src/analyses/constants.ml#L16-L17
https://github.com/goblint/analyzer/blob/262b6dab31938fafbc3cd9237f40b134812bd27d/src/analyses/constants.ml#L98

But the base analysis does the exact same thing:
https://github.com/goblint/analyzer/blob/262b6dab31938fafbc3cd9237f40b134812bd27d/src/cdomains/baseDomain.ml#L11-L12
https://github.com/goblint/analyzer/blob/262b6dab31938fafbc3cd9237f40b134812bd27d/src/analyses/base.ml#L51

So by the same logic our base analysis is wrong too because the same empty (`bot`) map would be if you just run base (without threadid and maybe others which always have non-bottom state that prevents this issue from occurring) on the same example as well.

---

Bottom of one analysis doesn't inherently mean dead code, `raise Deadcode` means dead code and that's handled by `DeadcodeLifter` and it's special bottom.
Just think about the unit analysis: it's only possible value is `()`, which is also the bottom of the unit lattice. So if you were to just activate the unit analysis, which is the stupidest possible thing, it's bottom everywhere. But that doesn't mean it's soundly analyzed all the code to be dead. The code is only considered dead when some analysis (hopefully soundly) raises `Deadcode` to explicitly indicate that.